clients/drcachesim/scheduler/scheduler.h

```diff
@@ -0,0 +1,562 @@
+        STATUS_ERROR_NOT_IMPLEMENTED,   /**< Error: not implemented. */
+    };
+
+    /**
+     * Status codes used by stream member functions.
+     * get_error_string() provides additional information such was which input file
+     * path failed.
+     */
+    enum stream_status_t {
+        STATUS_OK,      /**< Stream is healthy and can continue to advance. */
+        STATUS_EOF,     /**< Stream is at its end. */
+        STATUS_IDLE,    /**< No records at this time. */
+        STATUS_WAIT,    /**< Wait for other streams before advancing can continue. */
+        STATUS_INVALID, /**< Error condition. */
+        STATUS_REGION_INVALID,  /**< Input region is out of bounds. */
+        STATUS_NOT_IMPLEMENTED, /**< Feature not implemented. */
+    };
+
+    /** A bounded sequence of instructions. */
+    struct range_t {
+        /** Convenience constructor. */
+        range_t(uint64_t start, uint64_t stop)
+            : start_instruction(start)
+            , stop_instruction(stop)
+        {
+        }
+        /**
+         * The starting point as a trace instruction ordinal.  These ordinals
+         * begin at 1, so a 'start_instruction' value of 0 is invalid.
+         */
+        uint64_t start_instruction;
+        /** The ending point.  A stop value of 0 means the end of the trace. */
+        uint64_t stop_instruction;
+    };
+
+    /**
+     * Specifies details about one set of input trace files, each representing a
+     * software thread.
+     */
+    struct input_thread_info_t {
+        /** Convenience constructor for common usage. */
+        explicit input_thread_info_t(std::vector<range_t> regions)
+            : struct_size(sizeof(input_thread_info_t))
+            , regions_of_interest(regions)
+        {
+        }
+        size_t struct_size; /**< Size of the struct for binary-compatible additions. */
+        /**
+         * Which threads the details in this structure apply to.
+         * If empty, the details apply to all threads in the
+         * #dynamorio::drmemtrace::scheduler_tmpl_t::input_workload_t.
+         */
+        std::vector<memref_tid_t> tids;
+        /**
+         * Limits these threads to this set of output streams.  They will not
+         * be scheduled on any other output streams.
+         */
+        std::set<int> output_binding;
+        /**
+         * Relative priority for scheduling.  The default is 0.
+         */
+        // TODO i#5843: Decide and document whether these priorities are strict
+        // and higher will starve lower or whether they are looser.
+        int priority = 0;
+        /**
+         * If non-empty, all input records outside of these ranges are skipped: it is as
+         * though the input were constructed by concatenating these ranges together.  A
+         * #TRACE_MARKER_TYPE_WINDOW_ID marker is inserted between ranges (with a value
+         * equal to the range ordinal) to notify the client of the discontinuity (but
```

Comment:
Because it's not a discontinuity as the user never saw the initial records.  What is the use case you're thinking of that would want a window id there?  Just to know it's not the actual start: but what action would the user take?

Reply:
OK, if you both want a marker up front: done.

Reply:
Actually this ends up breaking several tests as the view tool is not prepared for a window id to be the very first thing it outputs.  I think this is causing more complexity than it is solving.  Going back to what I had.